File: tools/pm-ai-rdp-remote-launcher/RpaScenarioArgumentSupport.cs
namespace PmAi.RdpRemoteLauncher;

/// <summary>
/// RPA 起動引数内の {@code --scenario} パスを、空白トークン分割なしで抽出・正規化する。
/// </summary>
internal static class RpaScenarioArgumentSupport
{
    internal static string NormalizeScenarioArguments(string? arguments)
    {
        return RebuildScenarioArguments(arguments, stripEternalFlag: true);
    }

    internal static string RepairScenarioArguments(string? arguments)
    {
        return RebuildScenarioArguments(arguments, stripEternalFlag: false);
    }

    internal static IReadOnlyList<string> ExtractScenarioPaths(string? arguments)
    {
        if (string.IsNullOrWhiteSpace(arguments))
        {
            return Array.Empty<string>();
        }

        return ParseScenarioAndOtherArguments(arguments.Trim(), stripEternalFlag: false).ScenarioPaths;
    }

    private static string RebuildScenarioArguments(string? arguments, bool stripEternalFlag)
    {
        if (string.IsNullOrWhiteSpace(arguments))
        {
            return string.Empty;
        }

        var parsed = ParseScenarioAndOtherArguments(arguments.Trim(), stripEternalFlag);
        var normalized = new List<string>(parsed.OtherTokens);
        foreach (var path in parsed.ScenarioPaths)
        {
            normalized.Add(AladdinRpaLaunchArgs.ScenarioFlag);
            normalized.Add(UncPathSegmentRepair.Repair(path));
        }

        if (normalized.Count == 0)
        {
            return string.Empty;
        }

        return WindowsArgumentFormatter.FormatArgumentTokens(normalized);
    }

    private static ParsedArguments ParseScenarioAndOtherArguments(string input)
    {
        return ParseScenarioAndOtherArguments(input, stripEternalFlag: true);
    }

    private static ParsedArguments ParseScenarioAndOtherArguments(
        string input,
        bool stripEternalFlag)
    {
        var scenarioPaths = new List<string>();
        var otherTokens = new List<string>();
        var index = 0;
        while (index < input.Length)
        {
            index = SkipWhitespace(input, index);
            if (index >= input.Length)
            {
                break;
            }

            if (StartsWithFlag(input, index, AladdinRpaLaunchArgs.ScenarioFlag))
            {
                index += AladdinRpaLaunchArgs.ScenarioFlag.Length;
                index = SkipWhitespace(input, index);
                var extracted = ExtractScenarioPath(input, index);
                if (!string.IsNullOrWhiteSpace(extracted.Path))
                {
                    scenarioPaths.Add(extracted.Path);
                }

                index = extracted.EndIndex;
                continue;
            }

            if (TryExtractBareScenarioPath(input, index, out var barePath, out var bareEnd))
            {
                scenarioPaths.Add(barePath);
                index = bareEnd;
                continue;
            }

            if (StartsWithFlag(input, index, AladdinRpaLaunchArgs.IdFlag)
                || StartsWithFlag(input, index, AladdinRpaLaunchArgs.PasswordFlag))
            {
                index = SkipFlagWithValue(input, index);
                continue;
            }

            var tokenExtracted = ReadNextToken(input, index);
            if (!string.IsNullOrEmpty(tokenExtracted.Token))
            {
                if (stripEternalFlag
                    && string.Equals(
                        tokenExtracted.Token,
                        AladdinRpaLaunchArgs.EternalFlag,
                        StringComparison.OrdinalIgnoreCase))
                {
                    index = tokenExtracted.EndIndex;
                    continue;
                }

                otherTokens.Add(tokenExtracted.Token);
            }

            index = tokenExtracted.EndIndex;
        }

        return new ParsedArguments(scenarioPaths, otherTokens);
    }

    private static ExtractedPath ExtractScenarioPath(string input, int startIndex)
    {
        if (startIndex >= input.Length)
        {
            return new ExtractedPath(string.Empty, startIndex);
        }

        if (input[startIndex] == '"')
        {
            var end = FindClosingQuote(input, startIndex + 1);
            if (end < 0)
            {
                return new ExtractedPath(string.Empty, startIndex);
            }

            var path = input[(startIndex + 1)..end].Replace("\"\"", "\"");
            return new ExtractedPath(path, end + 1);
        }

        var ardrpaIndex = IndexOfIgnoreCase(input, ".ardrpa", startIndex);
        if (ardrpaIndex < 0)
        {
            var token = ReadNextToken(input, startIndex);
            return new ExtractedPath(token.Token, token.EndIndex);
        }

        var endIndex = ardrpaIndex + ".ardrpa".Length;
        return new ExtractedPath(input[startIndex..endIndex], endIndex);
    }

    private static bool TryExtractBareScenarioPath(
        string input,
        int startIndex,
        out string path,
        out int endIndex)
    {
        path = string.Empty;
        endIndex = startIndex;
        if (startIndex >= input.Length)
        {
            return false;
        }

        if (input[startIndex] == '"')
        {
            return false;
        }

        if (!LooksLikePathStart(input, startIndex))
        {
            return false;
        }

        var ardrpaIndex = IndexOfIgnoreCase(input, ".ardrpa", startIndex);
        if (ardrpaIndex < 0)
        {
            return false;
        }

        endIndex = ardrpaIndex + ".ardrpa".Length;
        path = input[startIndex..endIndex];
        return true;
    }

    private static bool LooksLikePathStart(string input, int startIndex)
    {
        if (startIndex + 1 < input.Length && input[startIndex] == '\\' && input[startIndex + 1] == '\\')
        {
            return true;
        }

        return startIndex + 2 < input.Length
               && char.IsLetter(input[startIndex])
               && input[startIndex + 1] == ':';
    }

    private static TokenExtract ReadNextToken(string input, int startIndex)
    {
        if (startIndex >= input.Length)
        {
            return new TokenExtract(string.Empty, startIndex);
        }

        if (input[startIndex] == '"')
        {
            var end = FindClosingQuote(input, startIndex + 1);
            if (end < 0)
            {
                return new TokenExtract(input[startIndex..], input.Length);
            }

            return new TokenExtract(
                input[(startIndex + 1)..end].Replace("\"\"", "\""),
                end + 1);
        }

        var index = startIndex;
        while (index < input.Length && !char.IsWhiteSpace(input[index]))
        {
            index++;
        }

        return new TokenExtract(input[startIndex..index], index);
    }

    private static int SkipFlagWithValue(string input, int startIndex)
    {
        var token = ReadNextToken(input, startIndex);
        var index = SkipWhitespace(input, token.EndIndex);
        if (index >= input.Length || input[index] == '-')
        {
            return index;
        }

        return ReadNextToken(input, index).EndIndex;
    }

    private static int SkipWhitespace(string input, int startIndex)
    {
        var index = startIndex;
        while (index < input.Length && char.IsWhiteSpace(input[index]))
        {
            index++;
        }

        return index;
    }

    private static bool StartsWithFlag(string input, int startIndex, string flag)
    {
        if (startIndex + flag.Length > input.Length)
        {
            return false;
        }

        return string.Compare(
            input,
            startIndex,
            flag,
            0,
            flag.Length,
            StringComparison.OrdinalIgnoreCase) == 0;
    }

    private static int FindClosingQuote(string text, int fromIndex)
    {
        for (var i = fromIndex; i < text.Length; i++)
        {
            if (text[i] != '"')
            {
                continue;
            }

            if (i + 1 < text.Length && text[i + 1] == '"')
            {
                i++;
                continue;
            }

            return i;
        }

        return -1;
    }

    private static int IndexOfIgnoreCase(string text, string value, int startIndex)
    {
        return text.IndexOf(value, startIndex, StringComparison.OrdinalIgnoreCase);
    }

    private readonly record struct ExtractedPath(string Path, int EndIndex);

    private readonly record struct TokenExtract(string Token, int EndIndex);

    private readonly record struct ParsedArguments(
        List<string> ScenarioPaths,
        List<string> OtherTokens);
}

File: code_java/src/main/java/jp/co/pm/ai/desktop/io/RpaScenarioArgumentSupport.java
package jp.co.pm.ai.desktop.io;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import jp.co.pm.ai.desktop.config.AladdinRpaLaunchArgs;

/**
 * RPA 起動引数内の {@link AladdinRpaLaunchArgs#SCENARIO_FLAG} パスを、空白トークン分割なしで抽出・正規化する。
 */
final class RpaScenarioArgumentSupport {

    private RpaScenarioArgumentSupport() {}

    static String normalizeScenarioArguments(String arguments) {
        return rebuildScenarioArguments(arguments, true);
    }

    static String repairScenarioArguments(String arguments) {
        return rebuildScenarioArguments(arguments, false);
    }

    static List<String> extractScenarioPaths(String arguments) {
        if (arguments == null || arguments.isBlank()) {
            return List.of();
        }
        return parseScenarioAndOtherArguments(arguments.trim(), false).scenarioPaths();
    }

    private static String rebuildScenarioArguments(String arguments, boolean stripEternalFlag) {
        if (arguments == null || arguments.isBlank()) {
            return "";
        }
        ParsedArguments parsed =
                parseScenarioAndOtherArguments(arguments.trim(), stripEternalFlag);
        List<String> normalized = new ArrayList<>(parsed.otherTokens());
        for (String path : parsed.scenarioPaths()) {
            normalized.add(AladdinRpaLaunchArgs.SCENARIO_FLAG);
            normalized.add(UncPathSegmentRepair.repair(path));
        }
        if (normalized.isEmpty()) {
            return "";
        }
        return formatTokensForIniArguments(normalized);
    }

    private static ParsedArguments parseScenarioAndOtherArguments(String input) {
        return parseScenarioAndOtherArguments(input, true);
    }

    private static ParsedArguments parseScenarioAndOtherArguments(
            String input, boolean stripEternalFlag) {
        List<String> scenarioPaths = new ArrayList<>();
        List<String> otherTokens = new ArrayList<>();
        int index = 0;
        while (index < input.length()) {
            index = skipWhitespace(input, index);
            if (index >= input.length()) {
                break;
            }
            if (startsWithFlag(input, index, AladdinRpaLaunchArgs.SCENARIO_FLAG)) {
                index += AladdinRpaLaunchArgs.SCENARIO_FLAG.length();
                index = skipWhitespace(input, index);
                ExtractedPath extracted = extractScenarioPath(input, index);
                if (!extracted.path().isBlank()) {
                    scenarioPaths.add(extracted.path());
                }
                index = extracted.endIndex();
                continue;
            }
            if (tryExtractBareScenarioPath(input, index) instanceof ExtractedPath bare
                    && !bare.path().isBlank()) {
                scenarioPaths.add(bare.path());
                index = bare.endIndex();
                continue;
            }
            if (startsWithFlag(input, index, AladdinRpaLaunchArgs.ID_FLAG)
                    || startsWithFlag(input, index, AladdinRpaLaunchArgs.PASSWORD_FLAG)) {
                index = skipFlagWithValue(input, index);
                continue;
            }
            TokenExtract token = readNextToken(input, index);
            if (!token.token().isEmpty()) {
                if (stripEternalFlag
                        && AladdinRpaLaunchArgs.ETERNAL_FLAG.equalsIgnoreCase(token.token())) {
                    index = token.endIndex();
                    continue;
                }
                otherTokens.add(token.token());
            }
            index = token.endIndex();
        }
        return new ParsedArguments(scenarioPaths, otherTokens);
    }

    private static ExtractedPath extractScenarioPath(String input, int startIndex) {
        if (startIndex >= input.length()) {
            return new ExtractedPath("", startIndex);
        }
        if (input.charAt(startIndex) == '"') {
            int end = findClosingQuote(input, startIndex + 1);
            if (end < 0) {
                return new ExtractedPath("", startIndex);
            }
            String path = input.substring(startIndex + 1, end).replace("\"\"", "\"");
            return new ExtractedPath(path, end + 1);
        }
        int ardrpaIndex = indexOfIgnoreCase(input, ".ardrpa", startIndex);
        if (ardrpaIndex < 0) {
            TokenExtract token = readNextToken(input, startIndex);
            return new ExtractedPath(token.token(), token.endIndex());
        }
        int endIndex = ardrpaIndex + ".ardrpa".length();
        return new ExtractedPath(input.substring(startIndex, endIndex), endIndex);
    }

    private static ExtractedPath tryExtractBareScenarioPath(String input, int startIndex) {
        if (startIndex >= input.length() || input.charAt(startIndex) == '"') {
            return null;
        }
        if (!looksLikePathStart(input, startIndex)) {
            return null;
        }
        int ardrpaIndex = indexOfIgnoreCase(input, ".ardrpa", startIndex);
        if (ardrpaIndex < 0) {
            return null;
        }
        int endIndex = ardrpaIndex + ".ardrpa".length();
        return new ExtractedPath(input.substring(startIndex, endIndex), endIndex);
    }

    private static boolean looksLikePathStart(String input, int startIndex) {
        if (startIndex + 1 < input.length()
                && input.charAt(startIndex) == '\\'
                && input.charAt(startIndex + 1) == '\\') {
            return true;
        }
        return startIndex + 2 < input.length()
                && Character.isLetter(input.charAt(startIndex))
                && input.charAt(startIndex + 1) == ':';
    }

    private static TokenExtract readNextToken(String input, int startIndex) {
        if (startIndex >= input.length()) {
            return new TokenExtract("", startIndex);
        }
        if (input.charAt(startIndex) == '"') {
            int end = findClosingQuote(input, startIndex + 1);
            if (end < 0) {
                return new TokenExtract(input.substring(startIndex), input.length());
            }
            return new TokenExtract(
                    input.substring(startIndex + 1, end).replace("\"\"", "\""), end + 1);
        }
        int index = startIndex;
        while (index < input.length() && !Character.isWhitespace(input.charAt(index))) {
            index++;
        }
        return new TokenExtract(input.substring(startIndex, index), index);
    }

    private static int skipFlagWithValue(String input, int startIndex) {
        TokenExtract token = readNextToken(input, startIndex);
        int index = skipWhitespace(input, token.endIndex());
        if (index >= input.length() || input.charAt(index) == '-') {
            return index;
        }
        return readNextToken(input, index).endIndex();
    }

    private static int skipWhitespace(String input, int startIndex) {
        int index = startIndex;
        while (index < input.length() && Character.isWhitespace(input.charAt(index))) {
            index++;
        }
        return index;
    }

    private static boolean startsWithFlag(String input, int startIndex, String flag) {
        if (startIndex + flag.length() > input.length()) {
            return false;
        }
        return input.regionMatches(true, startIndex, flag, 0, flag.length());
    }

    private static int findClosingQuote(String text, int fromIndex) {
        for (int i = fromIndex; i < text.length(); i++) {
            if (text.charAt(i) != '"') {
                continue;
            }
            if (i + 1 < text.length() && text.charAt(i + 1) == '"') {
                i++;
                continue;
            }
            return i;
        }
        return -1;
    }

    private static int indexOfIgnoreCase(String text, String value, int startIndex) {
        return text.toLowerCase(Locale.ROOT)
                .indexOf(value.toLowerCase(Locale.ROOT), startIndex);
    }

    private static String formatTokensForIniArguments(List<String> tokens) {
        StringBuilder out = new StringBuilder();
        for (String token : tokens) {
            if (token.isEmpty()) {
                continue;
            }
            if (out.length() > 0) {
                out.append(' ');
            }
            out.append(quoteArgumentIfNeeded(token));
        }
        return out.toString();
    }

    private static String quoteArgumentIfNeeded(String token) {
        if (token.indexOf(' ') >= 0 || token.indexOf('\t') >= 0) {
            return "\"" + token.replace("\"", "\"\"") + "\"";
        }
        return token;
    }

    private record ParsedArguments(List<String> scenarioPaths, List<String> otherTokens) {}

    private record ExtractedPath(String path, int endIndex) {}

    private record TokenExtract(String token, int endIndex) {}
}
